From a4b4502134b0c668977df68cd858400569209a16 Mon Sep 17 00:00:00 2001
From: Ding Yunliang <yunliang.ding@intel.com>
Date: Thu, 30 Jul 2015 10:03:03 +0800
Subject: [PATCH 458/462] ov8858: exchange s_power and update_setting

When calling ov8858 config, there will be a power up action
which will power up default VCM, then update setting. This
change will make sure power off default VCM called first to
avoid any power leakage.

Change-Id: I9716b0a1dbfd4db61e04962dad95c20d1dc5f98c
Signed-off-by: Ding Yunliang <yunliang.ding@intel.com>
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-14159
Reviewed-on: https://icggerrit.ir.intel.com/67606
Reviewed-by: Zaikuo Wang <zaikuo.wang@intel.com>
Validation-Android: Wei Tao <wei.tao@intel.com>
Reviewed-by: Jarno Jakonen <jarno.jakonen@intel.com>
Approver: Bingbu Cao <bingbu.cao@intel.com>
Reviewed-by: Bingbu Cao <bingbu.cao@intel.com>
Integrator: WeiX Meng <weix.meng@intel.com>
Build: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Maintainer: WeiX Meng <weix.meng@intel.com>
---
 .../camera/drivers/media/i2c/imx/dw9718.c          |  4 +---
 .../camera/drivers/media/i2c/ov8858.c              | 22 ++++++++++++++--------
 2 files changed, 15 insertions(+), 11 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c b/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c
index ec34e6e..65a1fcf 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c
@@ -87,10 +87,8 @@ int dw9718_t_focus_abs(struct v4l2_subdev *sd, s32 value)
 	value = clamp(value, 0, DW9718_MAX_FOCUS_POS);
 	ret = dw9718_i2c_wr16(client, DW9718_DATA_M, value);
 	/*pr_info("%s: value = %d\n", __func__, value);*/
-	if (ret < 0) {
-		dev_err(&client->dev, "write DW9718_DATA_M  failed %d\n", ret);
+	if (ret < 0)
 		return ret;
-	}
 
 	getnstimeofday(&dw9718_dev.focus_time);
 	dw9718_dev.focus = value;
diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov8858.c b/drivers/external_drivers/camera/drivers/media/i2c/ov8858.c
index b9c711d..d5d31a6 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov8858.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov8858.c
@@ -1699,24 +1699,30 @@ static int ov8858_s_config(struct v4l2_subdev *sd,
 
 	dev->sensor_id = sensor_id;
 
-	/* Resolution settings depend on sensor type and platform */
-	ret = __update_ov8858_device_settings(dev, dev->sensor_id);
-	if (ret)
-		goto fail_detect;
-
 	/* power off sensor */
 	ret = __ov8858_s_power(sd, 0);
+	if (ret) {
+		dev->platform_data->csi_cfg(sd, 0);
+		dev_err(&client->dev, "__ov8858_s_power-down error %d!\n", ret);
+		goto fail_update;
+	}
 
-	mutex_unlock(&dev->input_lock);
-	if (ret)
-		dev_err(&client->dev, "power-down error %d!\n", ret);
+	/* Resolution settings depend on sensor type and platform */
+	ret = __update_ov8858_device_settings(dev, dev->sensor_id);
+	if (ret) {
+		dev->platform_data->csi_cfg(sd, 0);
+		dev_err(&client->dev, "__update_ov8858_device_settings error %d!\n", ret);
+		goto fail_update;
+	}
 
+	mutex_unlock(&dev->input_lock);
 	return ret;
 
 fail_detect:
 	dev->platform_data->csi_cfg(sd, 0);
 fail_csi_cfg:
 	__ov8858_s_power(sd, 0);
+fail_update:
 	if (dev->platform_data->platform_deinit)
 		dev->platform_data->platform_deinit();
 	mutex_unlock(&dev->input_lock);
-- 
1.9.1

